projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,344 @@
+    received = datetime.fromtimestamp(p["meta/Timestamp"] / 1000.0 / 1000.0 / 1000.0)
+    sent = datetime.fromtimestamp(mktime_tz(parsedate_tz(p["meta/Date"]))) if p["meta/Date"] is not None else received
+    clock_skew = received - sent
+
+    return (received - created - clock_skew).total_seconds()
+```
+### Validate the ping payload
+Check that the payload section contains the right entries with consistent values.
+
+
+```python
+subset = get_pings_properties(update_pings, ["id",
+                                             "clientId",
+                                             "meta/creationTimestamp",
+                                             "meta/Date",
+                                             "meta/Timestamp",
+                                             "application/buildId",
+                                             "application/channel",
+                                             "application/version",
+                                             "environment/system/os/name",
+                                             "payload/reason",
+                                             "payload/targetBuildId",
+                                             "payload/targetChannel",
+                                             "payload/targetVersion",
+                                             "payload/previousBuildId",
+                                             "payload/previousChannel",
+                                             "payload/previousVersion"])
+```
+
+```python
+ping_success = subset.filter(lambda p: p.get("payload/reason") == "success")
+ping_ready = subset.filter(lambda p: p.get("payload/reason") == "ready")
+
+ping_success_count = ping_success.count()
+ping_ready_count = ping_ready.count()
+ping_count = ping_ready_count + ping_success_count
+```
+Quantify the percentage of duplicate pings we're receiving. We don't expect this value to be greater than ~1%, which is the amount we usually get from `main` and `crash`: as a rule of thumb, we threat anything less than 1% as *probably* well behaving.
+
+
+```python
+deduped_subset = dedupe(ping_success, "id")
+deduped_count = deduped_subset.count()
+print("Percentage of duplicate pings: {:.3f}".format(100.0 - pct(deduped_count, ping_success_count)))
+```
+    Percentage of duplicate pings: 0.078
+
+
+The percentage of duplicate pings is within the expected range. Move on and verify the payload of the `update` pings.
+
+
+```python
+def validate_update_payload(p):
+    PAYLOAD_KEYS = [
+        "payload/reason",
+        "payload/previousBuildId",
+        "payload/previousChannel",
+        "payload/previousVersion"
+    ]
+
+    # All the payload keys needs to be strings.
+    for k in PAYLOAD_KEYS:
+        if not isinstance(p.get(k), basestring):
+            return ("'{}' is not a string".format(k), 1)
+        
+    # We only expect "reason" = success.
+    if p.get("payload/reason") != "success":
```

Comment:
You filtered on this above in `ping_success`. Did you mean, in line 139, to dedupe `subset` instead of `ping_success`

Reply:
Yeah, this is a bogus check, removing it.